04-frequent-scenarios.md

```diff
@@ -0,0 +1,288 @@
+
+Ideally, if a resource has been moved to another link, the client will be told so in the standard HTTP way. In this case, the server method responsible for looking up the resource will throw an exception that is handled by a @ControllerAdvice method. This method will set the response code (301) and add the header with the new location of the resource.
+
+```java
+@ExceptionHandler(CarRedirectException.class)
+@ResponseStatus(HttpStatus.MOVED_PERMANENTLY)
+public void handleCarRedirect(CarRedirectException ex) {
+    resp.setHeader("Location", ex.getRedirectLink());
+}
+```
+
+No client-side measures need to be taken to deal with the redirect. The browser takes care of this handling.
+
+### 304 / Apply cache
+A cache is by default implemented in every browser. It will cache resources that a user visited. When those resource are once more requested, the cache will serve the cached resources if they are still 'fresh'. The most important aspect for a server to implement is the mechanism to determine 'freshness' of a resource.
+
+The following cache control mechanisms are supported by the browser cache:
+* do not cache; also known as 'no-store'. This forbids your browser to cache at all.
+* use cache, but verify first; also known as 'no-cache' (don't ask). The browser will first fetch the resource from the server, passing one of two, or both optional headers (If-Modified-Since, If-None-Match). If the servers returns a 304 status code, it will use its cached resource. Otherwise, it will use the response and cache it.
+* use cache, but verify first if resource is stale; also known as 'must-revalidate'. Works similar to 'no-cache', but only calls the server if the maximum age has not been exceeded. Additionaly, on calling the server, the age will be updated if the resource is still fresh.
+
+For regular browser caches, the 'no-cache' option is preferred. Passing a max age, means it will be hard to evict a resource from cache until it has grown stale. The 'no-cache' has no such problems.
+
+A case can be made to use 'must-revalidate' in combination with a 'max-age', especially for sites that expect high traffic load on resources that are consumed manifold, but updated rarely.
+
+Request
+```
+GET /cars/1
+If-Modified-Since: [DATE]
+If-None-Match: [ETAG]
+```
+
+Response
+```
+Status code: 304
+Cache-Control: [CACHE INSTRUCTIONS]
+Last-Modified: [DATE] 
+Etag: [ETAG]
+```
+
+No special measures have to be made in the browser application. The browser cache will automatically take care of the response and all required actions to update the cache.
+
+The values for the Last-Modified and/or Etag are determined by the server. The client will use those values literally.
```

Comment:
Etag is meestal een hash van de content. Dit is de conventie, maar niet verplicht. Het belangrijkste van de Etag is dat over de content consistent dezelfde hash gegenereerd wordt zodat bepaald kan worden of de content veranderd is.